bigquery_etl/backfill/shredder_mitigation.py

```diff
@@ -50,30 +67,187 @@ class DataTypeGroup(Enum):
+            return None
+        return int(self.expiration_days * 86400000)
+
+    @property
+    def name(self):
+        if self.suffix is not None and self.suffix != "":
+            return f"{self.destination_table}{self.suffix}"
+        else:
+            return self.destination_table
+
+    @property
+    def version(self):
+        match = re.search(r"v(\d+)$", self.destination_table)
+        return int(match.group(1))
+
+    @property
+    def name_previous_version(self):
+        """Return the table name for the previous version of the table."""
+        return f"{self.destination_table[:-1]}{int(self.version)-1}"
+
+    @property
+    def full_table_id(self):
+        return f"{self.project_id}.{self.dataset}.{self.name}"
+
+    @property
+    def partitioning(self) -> str:
+        metadata = Metadata.from_file(
+            Path("sql") / self.project_id / self.dataset / self.name / METADATA_FILE
         )
+        if metadata.bigquery and metadata.bigquery.time_partitioning:
+            partitioning = {
+                "type": metadata.bigquery.time_partitioning.type,
+                "field": metadata.bigquery.time_partitioning.field,
+            }
+        return partitioning
+
+    def generate_query(
+        self,
+        select_list,
+        from_clause=None,
+        where_clause=None,
+        group_by_clause=None,
+        order_by_clause=None,
+    ):
+        """Build query to populate the table."""
+
+        query = f"SELECT {', '.join(select_list)}"
+        query += f" {from_clause}" if from_clause is not None else ""
+        query += f" {where_clause}" if where_clause is not None else ""
+        query += f" {group_by_clause}" if group_by_clause is not None else ""
+        query += f" {order_by_clause};" if order_by_clause is not None else ""
+        return query
+
+    def get_query_path(self):
+        """Return the path of the query file for this subset."""
+        query_files = paths_matching_name_pattern(
+            f"{self.dataset}.{self.name}", "sql", self.project_id
+        )
+        if query_files == []:
+            raise click.ClickException(
+                f"Query file not found for `{self.full_table_id}`."
+            )
+        return query_files[0]
+
+    def get_query_path_results(
+        self, backfill_date: date = PREVIOUS_DATE, rows: int = None, **kwargs
+    ) -> list[{}]:
+        """Run the query available in the sql_path and return the result or the number of rows requested."""
+
+        having_clause = None
+        for key, value in kwargs.items():
+            if key.lower() == "having_clause":
+                having_clause = f"{value}"
+
+        with open(self.get_query_path(), "r") as file:
+            sql_text = file.read().strip()
+
+        if sql_text.endswith(";"):
+            sql_text = sql_text[:-1]
+        if having_clause:
+            sql_text = f"{sql_text} {having_clause}"
+        if rows:
+            sql_text = f"{sql_text} LIMIT {rows};"
```

Comment:
That's me wanting the queries to be tidy and end with ; in case there is another query after. In this case it's not required so instead of moving it outside LIMIT, it's simpler to not add it. Thanks,